main/src/ca/uwaterloo/flix/language/phase/Reducer.scala

```diff
@@ -33,131 +33,152 @@ import scala.jdk.CollectionConverters._
-
-    case Expr.Branch(exp, branches, tpe, purity, loc) =>
-      val e = visitExpr(exp)
-      val bs = branches map {
-        case (label, body) => label -> visitExpr(body)
+      // Compute the types in the captured formal parameters.
+      val cParamTypes = cparams.foldLeft(Set.empty[MonoType]) {
+        case (sacc, FormalParam(_, _, tpe, _)) => sacc + tpe
       }
-      Expr.Branch(e, bs, tpe, purity, loc)
-
-    case Expr.JumpTo(sym, tpe, purity, loc) =>
-      Expr.JumpTo(sym, tpe, purity, loc)
-
-    case Expr.Let(sym, exp1, exp2, tpe, purity, loc) =>
-      lctx.lparams.addOne(LocalParam(sym, exp1.tpe))
-      val e1 = visitExpr(exp1)
-      val e2 = visitExpr(exp2)
-      Expr.Let(sym, e1, e2, tpe, purity, loc)
-
-    case Expr.LetRec(varSym, index, defSym, exp1, exp2, tpe, purity, loc) =>
-      lctx.lparams.addOne(LocalParam(varSym, exp1.tpe))
-      val e1 = visitExpr(exp1)
-      val e2 = visitExpr(exp2)
-      Expr.LetRec(varSym, index, defSym, e1, e2, tpe, purity, loc)
-
-    case Expr.Stmt(exp1, exp2, tpe, purity, loc) =>
-      val e1 = visitExpr(exp1)
-      val e2 = visitExpr(exp2)
-      Expr.Stmt(e1, e2, tpe, purity, loc)
-
-    case Expr.Scope(sym, exp, tpe, purity, loc) =>
-      lctx.lparams.addOne(LocalParam(sym, MonoType.Region))
-      val e = visitExpr(exp)
-      Expr.Scope(sym, e, tpe, purity, loc)
 
-    case Expr.TryCatch(exp, rules, tpe, purity, loc) =>
-      val e = visitExpr(exp)
-      val rs = rules map {
-        case CatchRule(sym, clazz, body) =>
-          lctx.lparams.addOne(LocalParam(sym, MonoType.Object))
-          val b = visitExpr(body)
-          CatchRule(sym, clazz, b)
-      }
-      Expr.TryCatch(e, rs, tpe, purity, loc)
+      // `defn.fparams` and `defn.tpe` are both included in `defn.arrowType`
+      ctx.defTypes.add(d.arrowType)
+      ctx.defTypes.add(unboxedType.tpe)
+      ctx.defTypes.addAll(cParamTypes.asJavaCollection)
 
-    case Expr.TryWith(exp, effUse, rules, tpe, purity, loc) =>
-      lctx.pcPoints += 1
-      val e = visitExpr(exp)
-      val rs = rules.map {
-        case HandlerRule(op, fparams, exp) =>
-          val e = visitExpr(exp)
-          HandlerRule(op, fparams, e)
-      }
-      Expr.TryWith(e, effUse, rs, tpe, purity, loc)
-
-    case Expr.Do(op, exps, tpe, purity, loc) =>
-      lctx.pcPoints += 1
-      val es = exps.map(visitExpr)
-      Expr.Do(op, es, tpe, purity, loc)
-
-    case Expr.NewObject(name, clazz, tpe, purity, methods, loc) =>
-      val specs = methods.map {
-        case JvmMethod(ident, fparams, clo, retTpe, purity, loc) =>
-          val c = visitExpr(clo)
-          JvmMethod(ident, fparams, c, retTpe, purity, loc)
-      }
-      ctx.anonClasses.add(AnonClass(name, clazz, tpe, specs, loc))
+      Def(ann, mod, sym, cparams, fparams, ls, pcPoints, e, tpe, unboxedType, purity, loc)
+  }
+
+  private def visitExpr(exp0: Expr)(implicit lctx: LocalContext, ctx: SharedContext): Expr = {
+    ctx.defTypes.add(exp0.tpe)
```

Comment:
All the cases that have `ctx.defTypes.add(tpe)` seems redundant because of this line right? You do like the code did previously but i think this should be the only call to ctx.detTypes.add

Reply:
A little over 30ms with this